<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [rabbitmq-discuss] it throws AlreadyClosedException, not IOException.
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20it%20throws%20AlreadyClosedException%2C%0A%20not%20IOException.&In-Reply-To=%3C596AA34C9AE73245BFEC2A5202A00BA1069FB87505%40EXCH-MBX-21.vmware.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="012583.html">
   <LINK REL="Next"  HREF="012562.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[rabbitmq-discuss] it throws AlreadyClosedException, not IOException.</H1>
    <B>Ann Witbrock</B> 
    <A HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20it%20throws%20AlreadyClosedException%2C%0A%20not%20IOException.&In-Reply-To=%3C596AA34C9AE73245BFEC2A5202A00BA1069FB87505%40EXCH-MBX-21.vmware.com%3E"
       TITLE="[rabbitmq-discuss] it throws AlreadyClosedException, not IOException.">awitbrock at vmware.com
       </A><BR>
    <I>Thu Apr 28 13:11:32 BST 2011</I>
    <P><UL>
        <LI>Previous message: <A HREF="012583.html">[rabbitmq-discuss]  it throws AlreadyClosedException, not IOException.
</A></li>
        <LI>Next message: <A HREF="012562.html">[rabbitmq-discuss] Message Aggregating Queue
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#12589">[ date ]</a>
              <a href="thread.html#12589">[ thread ]</a>
              <a href="subject.html#12589">[ subject ]</a>
              <a href="author.html#12589">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Having looked at the API docs, you might also expect to find a ShutdownSignalException.

Ann Witbrock
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">awitbrock at vmware.com</A>
________________________________________
From: <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss-bounces at lists.rabbitmq.com</A> [<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss-bounces at lists.rabbitmq.com</A>] On Behalf Of Ann Witbrock [<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">awitbrock at vmware.com</A>]
Sent: 28 April 2011 11:52
To: <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>
Subject: [rabbitmq-discuss]  it throws AlreadyClosedException, not IOException.

Closing the connection will also close the channel it created using connection.createChannel().
Therefore, calling a channel method after the connection is closed, should also find the channel closed, or closing.
This should raise an AlreadyClosedException, as you discovered, allowing you to know the connection and channel are absent, rather than performing unexpectedly.

Ann Witbrock
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">awitbrock at vmware.com</A>
________________________________________
From: <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss-bounces at lists.rabbitmq.com</A> [<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss-bounces at lists.rabbitmq.com</A>] On Behalf Of &#24352;&#21338;&#24378; [<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">boqiang.zhang at gmail.com</A>]
Sent: 28 April 2011 08:57
To: <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>
Subject: [rabbitmq-discuss] it throws AlreadyClosedException,   not IOException.

I'm using rabbitmq java client 2.4.1

After a TCP connection lost, and still call a method on a channel over
this connection, a AlreadyClosedException will be thrown.

it is a bug? I expected an IOException,  but AlreadyClosedException I
got, and AlreadyClosedException is a RuntimeException.

if not, why all other errors cause an IOException.

      @Test
      public void testConnectionLost() throws IOException{
                ConnectionFactory factory = new ConnectionFactory();
                factory.setRequestedHeartbeat(60);
                factory.setHost(&quot;&lt;your rabbitmq host&gt;&quot;);


                Connection connection = factory.newConnection();
                Channel channel = connection.createChannel();
                connection.close();

                try {
                        channel.queueDeclare(&quot;queueName&quot;, false, false, false, null);
                        Assert.fail(&quot;Exception expected.&quot;);
                }catch (IOException e) {
                        //it will NOT reach here.
                        //Inner exception should be AlreadyClosedException
                        System.out.println(e);
                }catch (AlreadyClosedException e) {
                        // it will reach here.
                        System.out.println(e);

                        //this is strange!
                        //I expected IOException , but AlreadyClosedException I got.
                        //And AlreadyClosedException is a RuntimeException.
                }

Thank you.
_______________________________________________
rabbitmq-discuss mailing list
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss</A>
_______________________________________________
rabbitmq-discuss mailing list
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>
<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss</A>
</PRE>













































































<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="012583.html">[rabbitmq-discuss]  it throws AlreadyClosedException, not IOException.
</A></li>
	<LI>Next message: <A HREF="012562.html">[rabbitmq-discuss] Message Aggregating Queue
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#12589">[ date ]</a>
              <a href="thread.html#12589">[ thread ]</a>
              <a href="subject.html#12589">[ subject ]</a>
              <a href="author.html#12589">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">More information about the rabbitmq-discuss
mailing list</a><br>
</body></html>
